lib/mix/lib/mix/scm/git.ex

```diff
@@ -123,7 +124,7 @@ defmodule Mix.SCM.Git do
   end
 
   defp do_checkout(opts) do
-    ref = get_lock_rev(opts[:lock]) || get_opts_rev(opts)
```

Comment:
To recap what we have discussed, I think the issue is all here. We should not trust the lock if the lock options are out of date, so this should rather be something like:

    ref =  get_lock_rev_if_recent(...) || get_opts_rev(opts)

This should make the rev be updated in the scenario where the bug was reported.